src/kakarot/instructions.cairo

```diff
@@ -34,80 +34,579 @@ from kakarot.instructions.sha3 import Sha3
 // @custom:namespace EVMInstructions
 namespace EVMInstructions {
     // @notice Decode the current opcode and execute associated function.
-    // @dev The function iterates through the provided instructions and executes each of them
-    //      whilst also performing safety checks and updating the pc counter after each instruction execution
-    // @param instructions The instruction set.
+    // @dev The function uses an internal jump table to execute the corresponding opcode
     // @param ctx The pointer to the execution context.
     // @return The pointer to the updated execution context.
-
     func decode_and_execute{
         syscall_ptr: felt*,
         pedersen_ptr: HashBuiltin*,
-        range_check_ptr,
+        range_check_ptr: felt,
         bitwise_ptr: BitwiseBuiltin*,
-    }(instructions: felt*, ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
         alloc_locals;
 
         // Retrieve the current program counter.
         let pc = ctx.program_counter;
-
-        // Revert if pc < 0
-        with_attr error_message("Kakarot: InvalidCodeOffset") {
-            assert_nn(pc);
-        }
-
         local opcode;
 
-        let is_pc_le_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
+        let is_pc_ge_code_len = is_le_felt(ctx.call_context.bytecode_len, pc);
```

Comment:
is_felt_le is not that cheap. Use the 128bit version if possible.

Reply:
good to know, thx